### ISSUE TITLE
Restrictions in TransformerLayer

### ISSUE THREAD

**Issue by rmanor** [Lasagne/Lasagne#440]
Hi,

Are there any plans to add optional restrictions to the TransformerLayer?
e.g., do only translation and scaling in one axis, etc.
Thanks.

**Comment by benanne** [comment]
I don't know if there are plans to provide explicit support for it, but you could simply add a custom layer that adequately restrict the 6 numbers that describe the transformation matrix. Maybe providing a few helper functions/layers for common cases (only rotations, only translations etc.) could be useful.

**Comment by skaae** [comment]
I don't think we plan to support that. 

@benanne: Is it easy to clamp outputs to a specific value? Maybe we could provide a layer for that?

**Comment by benanne** [comment]
Well, it's easy enough to write a custom layer. For example, you could have a layer that takes 1 single value as input (a rotation angle) and outputs the corresponding 6 numbers describing the transformation matrix. All you need is a class with two methods, both of which are probably 1-2 lines long. I'd say that's already easy enough and we don't need to complicate the `TransformerLayer` code to support this.

**Comment by skaae** [comment]
> I'd say that's already easy enough and we don't need to complicate the TransformerLayer code to support this.

agreed. I thought we could create a KeepConstantLayer that could clamp the outputs of the localization network. 
Something like
```Python
loc_constant = KeepConstant(loc, indices=slice(0,3))
transformer  = TransformerLayer(incoming=img, localization_network=loc_constant)
```

**Comment by benanne** [comment]
That seems like a bit of hacky solution though? It doesn't really cover all the bases. You don't necessarily want to keep certain values constant, sometimes the constraints take more complex forms (as in the case of rotations). So I feel like it makes more sense to have a neural net that outputs the parameters of whatever restricted set of transformations you've chosen, and then a simple custom layer that maps these onto the 6 values that describe general affine transformations.

**Comment by skaae** [comment]
ok. I havent really looked at how you constrain to e.g. rotations.

**Comment by ebenolson** [comment]
There seem to be a number of cases where you'd like to just tweak the output of a layer in some small way. What about adding something like this?

    class LambdaLayer(lasagne.layers.Layer):
        def __init__(self, incoming, function, output_shape=None, **kwargs):
            super(LambdaLayer, self).__init__(incoming, **kwargs)

            if output_shape is not None:
                if hasattr(output_shape, '__call__'):
                    self.get_output_shape_for = output_shape
                else:
                    self.input_shape = output_shape
            
            self.function = function
        
        def get_output_for(self, input, **kwargs):
            return self.function(input)

Then in this case you could just write

    loc_constant = LambdaLayer(loc, lambda X: X[slice(0, 3)], output_shape=(2,))

**Comment by skaae** [comment]
Good idea

**Comment by benanne** [comment]
I like the idea, but `self.input_shape = output_shape` seems pretty messy. I'm sure it works fine but `self.input_shape` no longer represents the actual input shape of the layer this way.

**Comment by ebenolson** [comment]
yeah, that was just a quick hack to see if it would work. I fleshed it out a little and made a PR for further discussion: https://github.com/Lasagne/Lasagne/pull/441
